Make TensorFlow use system/EasyBuild installed libraries

Flamefire opened the pull request:
TensorFlow allows using already installed libs. This avoids problems with the bundled TF libs (e.g. in 2.3.0) and speeds up the build process.

closes #1996

edit: now requires ~~https://github.com/easybuilders/easybuild-framework/pull/3426~~

boegel commented:
Tested with:

* [x] `TensorFlow-1.13.1-foss-2019a-Python-3.7.2.eb`
* [ ] `TensorFlow-1.15.2-fosscuda-2019b-Python-3.7.4.eb`
* [x] `TensorFlow-2.1.0-foss-2019b-Python-3.7.4.eb`
* [ ] `TensorFlow-2.2.0-foss-2019b-Python-3.7.4.eb`
* [ ] `TensorFlow-2.2.0-fosscuda-2019b-Python-3.7.4.eb`

Flamefire commented:
Thanks for the testing. This change should currently have no effect as the dependencies are missing. I added them to the 2.1.0 ECs in https://github.com/easybuilders/easybuild-easyconfigs/pull/11109 so you can add those to your tests too if you like.

Flamefire commented:
Also another idea: As TF likes to add/remove/rename packages, maybe we should add all names to the EasyBlock and verify the list. Even when we don't have a corresponding EC yet.

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +152,87 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+            ('NASM', None, ('2.1.0',)),
+            ('SQLite', 'org_sqlite', ('2.1.0',)),
+            ('PCRE', None, ('2.1.0',)),
+            ('libpng', 'png', ('2.1.0',)),
```
for our own sanity, we should keep this sorted alphabetically?

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +152,87 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+            ('gast', 'gast_archive', ('2.1.0',)),
+            ('Keras-Applications', 'keras_applications_archive', ('2.1.0', '2.2.0')),
+            ('opt-einsum', 'opt_einsum_archive', ('2.1.0',)),
+            ('google-pasta', 'pasta', ('2.1.0',)),
```
same here, please sort alphabetically

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +152,87 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+            ('termcolor', 'termcolor_archive', ('2.1.0',)),
+        )
+
+        dependencies = set(dep['name'] for dep in self.cfg.dependencies())
```
`dependencies` -> `dep_names`

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +152,87 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+        system_libs = []
+        cpaths = []
+        libpaths = []
+        for dep_name, tf_dep_name, version_requirement in available_system_libs:
```
rename `version_requirement` to `version_range`?

Flamefire replied:
I decided to keep it like that as it is not necessarily a range but can be a single element tuple (hopefully almost always)

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +152,87 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+        for dep_name, tf_dep_name, version_requirement in available_system_libs:
+            if dep_name in dependencies and is_version_ok(version_requirement):
+                system_libs.append(make_tf_name(dep_name, tf_dep_name))
+                sw_root = get_software_root(dep_name)
```
We should probably verify here if this is different from `None`, since due some dependencies may be being filtered (via `--filter-deps`). See also https://github.com/easybuilders/easybuild-framework/issues/1494

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +152,87 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+                if libpath:
+                    libpaths.append(os.path.join(sw_root, libpath))
+
+        for dep_name, tf_dep_name, version_requirement in python_system_libs:
```
also rename `version_requirement` to `version_range` (if you do so above)

Flamefire commented:
I addressed your comments and also some changes:

- introduce function get_system_libs_for_version which does the version filtering and can be used to render a list that can be diffed against the TF included list on updates. IMO crucial to keep up to date. We might even want to do this automatically during the build, what do you think?
- Add ALL TF known libs to a) allow for the above and b) see which are missing to potentially add new ECs or find the correct ones
- Combined the 2 lists into 1 and use values of `None` and `True` to denote missing ECs and "assume installed"
- Change order of tf_name and eb_name and sort lines. There are benefits in doing it this way (e.g. check against the tf list) but maybe the other way round is better as it allows to see name changes more easily as EC names stay (usually) constant (see jpeg). I tend to revert the order again but I'm still not 100% happy with the list anyway. Not great to read. Any ideas?

Flamefire commented:
Ok, still need to do a few adjustments. I'm adding and testing more ECs to be used as dependencies. To check I added a log which dependencies that could be used by TF are not.

The main dependency I'd like to include is protobuf but that needs another patch and a test. Checking now if that works...

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +153,135 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+            ('jpeg', 'libjpeg-turbo', ('2.1.0', '2.2.0')),
+            ('jsoncpp_git', 'JsonCpp', ('2.1.0',)),
+            ('keras_applications_archive', True, ('2.1.0', '2.2.0')),
+            ('libjpeg_turbo', 'libjpeg-turbo', ('2.2.0', )),
```
@Flamefire Ordering by EasyBuild software name, like it was before, makes more sense to me imho, since we're mapping from EB name (key) to TF name (value).

Maybe it should even be a proper dictionary (with dict values), where you filter the values based on the version range?

```python
available_system_libs = {
    'libjpeg-turbo': {
        ('2.1.0', '2.2.0'): 'jpeg',
        ('2.2.0',): 'libjpeg-turbo',
    },
    ...
    'SWIG': {('2.1.0',): 'swig'},
}
```

or even:

```python
available_system_libs = {
    'libjpeg-turbo': {
        '2.1.0:2.2.0': 'jpeg',
        '2.2.0:': 'libjpeg-turbo',
    },
    ...
    'SWIG': {'2.1.0:': 'swig'},
}
```

Flamefire replied:
The dict won't work due to the `True` and `None` values :/ And I'd like to keep them

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +153,135 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
             self.log.info("Using wrapper script for '%s': %s", compiler, which(compiler))
 
+    @staticmethod
+    def get_system_libs_for_version(tf_version, as_VALID_LIBS=False):
```
`as_VALID_LIBS` -> `as_valid_libs`

Flamefire replied:
That is intentional: In TF the name of the variable is `VALID_LIBS` and the format of the output is the same as that constant so you can diff them

Flamefire commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +278,103 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+                    dep_name = 'protobuf'
+                sw_root = get_software_root(dep_name)
+                # Dependency might be filtered via --filter-deps. In that case assume globally installed version
+                if not sw_root:
```
This code is only executed if `dep_name is in self.ec.dependencies()` and you told me to add the check for a defined `sw_root` to cater for this bug. So IMO all is handled, isn't it?

boegel replied:
Yes, I overlooked that this was already using `self.ec.dependencies()` higher up

Flamefire commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +278,103 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+                    if dep_name == 'protobuf':
+                        # Need to set INCLUDEDIR as TF wants to symlink headers from there:
+                        # https://github.com/tensorflow/tensorflow/issues/37835
+                        env.setvar('INCLUDEDIR', incpath)
```
> is that what we want to do?

Unfortunately yes. With a patch for TF I have `INCLUDEDIR` is only used by the Bazel code for protobuf (without it IIRC JsonCpp does use it too) so it is fine.

> Wouldn't it be more future-proof to take into account the existing $INCLUDEDIR value, 

The point is: And do what? It is an environment variable which TF expects to be set to 1 (and only 1) folder so there is nothing we could do with an existing value. The way they wrote that stuff is that ALL software is installed into 1 prefix, e.g. `/usr` -.-

boegel replied:
Right ok... It's not a `$PATH`-like variable, so little else we can do to overwrite it.

Flamefire commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +278,103 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+
+        # Check direct dependencies
+        dep_names = set(dep['name'] for dep in self.cfg.dependencies())
+        for dep_name, tf_name in dependency_mapping.items():
```
IIRC we are doing similar things at other places. I don't think there is any harm in the "unspecified" order, or do you see any issues?

boegel replied:
No real issues, it's just easier for humans staring at sorted things rather than having a random order (which may change between runs...).

Flamefire commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -237,6 +462,7 @@ def configure_step(self):
             'TF_NEED_KAFKA': '0',  # Amazon Kafka Platform
             'TF_SET_ANDROID_WORKSPACE': '0',
             'TF_DOWNLOAD_CLANG': '0',  # Still experimental in TF 2.1.0
+            'TF_SYSTEM_LIBS': ','.join(self.system_libs_info[0]),
```
I see no harm in being explicit here in NOT using any $TF_SYSTEM_LIBS. It will appear in the log as setting that to empty which is good IMO

boegel replied:
You're assuming here that every TensorFlow version won't trip over having `$TF_SYSTEM_LIBS` set to an empty value, which is technically not the same as not setting it (although in practice there's little difference).

Flamefire replied:
I'd argue that this assumption is safe. In fact current TF code contains `syslibs = environ_cp.get('TF_SYSTEM_LIBS', '')` so for that not setting it and setting it to empty is exactly the same. I verified that in 1.10.0 where the variable first appears the behavior is the same: empty and unset are treated equal

Therefore I'd like to go with the current solution which is cleaner in code (1 line less, no repetition of the dict var name) and more explicit (shows in the log as being set to empty)

boegel replied:
OK!

Flamefire commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +278,103 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+        self.log.info("Running command '%s'" % full_cmd)
+        proc = subprocess_popen_text(cmd_list, env=os.environ)
+        (stdout, stderr) = proc.communicate()
+        ec = proc.returncode
```
As discussed in slack this is postponed until implemented in framework

boegel replied:
Agreed.

Flamefire commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +278,103 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+            if build_option('strict') == run.ERROR:
+                raise EasyBuildError(msg)
+            else:
+                print_warning(msg)
```
I added `The EasyBlock needs to be updated to fully work with TF version %s` to the error and some comments above

Flamefire commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +278,103 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+        ec = proc.returncode
+        self.log.info("Command '%s' returned with %s: stdout: %s; stderr: %s" % (full_cmd, ec, stdout, stderr))
+        if ec:
+            raise EasyBuildError('Failed to determine installed python packages')
```
Added to the end

Flamefire commented:
@boegel needed to rebase to include the ninja fix to test this. New commits are starting at "Address review"

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +280,106 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+        if not available_libs_eb:
+            self.log.warning('TensorFlow EasyBlock does not have any information for $TF_SYSTEM_LIBS stored. ' +
+                             'This means most dependencies will be downloaded at build time by TensorFlow.\n' +
+                             'Available $TF_SYSTEM_LIBS according to the TF sources: %s', sorted(available_libs_src))
```
@Flamefire `TF sources` -> `TensorFlow sources` (it's fine to use `TF` in comments, but I wouldn't do that in warnings/errors)

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -152,6 +280,106 @@ def write_wrapper(self, wrapper_dir, compiler, i_mpi_root):
+                if unknown_libs:
+                    # Libs listed in this EasyBlock but not present in the TF sources -> Removed?
+                    msg += 'Unrecognized entries for $TF_SYSTEM_LIBS: %s\n' % unknown_libs
+                msg += 'The EasyBlock needs to be updated to fully work with TF version %s' % self.version
```
`TF version` -> `TensorFlow version`

boegel commented on easybuild/easyblocks/t/tensorflow.py:
```diff
@@ -73,6 +76,130 @@
+        txt = read_file(syslibs_path)
+        m = re.search(r'VALID_LIBS = \[(.*?)\]', txt, re.DOTALL)
+        if not m:
+            raise EasyBuildError('VALID_LIBS definition not found in %s', syslibs_path)
```
@Flamefire This is too strict, it breaks the installation of `TensorFlow-1.10.0-fosscuda-2018b-Python-2.7.15.eb` for example.

It seems OK for 1.13.1 and newer, so I would only error out for sufficiently recent versions?

Also, we shouldn't use a single-letter variable outside of a list comprehension, so please fix that too while you're at it (`valid_libs = re.search(...)`)

Flamefire replied:
Found that 1.10.0 (where this was introduced) is missing spaces. Adapted the regex.

It is actually only done for "sufficiently recent versions" where I defined this as "has the `syslibs_configure.bzl` file", which should be ok enough.

Fixed in https://github.com/easybuilders/easybuild-easyblocks/pull/2117/commits/969fcf245727d88b7dd7cb1cfd7154e380e98f22

boegel commented:
Doing a last round of testing, but this should be good to go now... Thanks a lot for your efforts on this @Flamefire!

boegel commented:
This has been tested extensively, both with a wide range of existing TensorFlow easyconfigs and the updated ones for TensorFlow 2.1.0 from https://github.com/easybuilders/easybuild-easyconfigs/pull/11109 that require these changes, so going in!
